feat(cache): messages support

noaccOS opened the pull request:
handle http and websocket messages cache, giving access to the original
controller (via its pid) or websocket (also its pid)

depends on #6 and includes its commits
closes #4

noaccOS committed:
chore: add elixir_uuid dependency

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
chore: add edgehog_device_forwarder dependency

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS commented:
Split the type `message_id` into `http_request_id` and `ws_socket_id` for clarity

rbino commented on lib/edgehog_device_forwarder/connections.ex:
```diff
@@ -0,0 +1,147 @@
+
+    {ConCache, params}
+    |> Supervisor.child_spec(id: id)
+  end
```
Ok now I get why you were using a Supervisor here, but you're stuffing everything in a single module for no reason. In this module there are 3 separate things:
- A Supervisor that starts the token and messages caches
- The API for the token cache
- The API for the message cache

These should be split in separate modules since they are 3 different components.

noaccOS replied:
Done

rbino commented on test/support/case.ex:
```diff
@@ -0,0 +1,30 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Case do
```
`Case` is too generic, usually you have stuff like `DataCase`, `ConnectionCase` etc. Think about what this case is covering specifically and name it like that.

rbino commented on lib/edgehog_device_forwarder/connections.ex:
```diff
@@ -0,0 +1,147 @@
+
+    case ConCache.insert_new(@messages, request_id, {:http, controller}) do
+      :ok -> {:ok, request_id}
+      {:error, :already_exists} -> new_http_request(controller)
```
Is that infinite recursion or am I missing something?

noaccOS replied:
it stops when it generates a new UUIDv4 (not already in the cache)

rbino commented on lib/edgehog_device_forwarder/connections.ex:
```diff
@@ -0,0 +1,147 @@
+  ## Messages
+
+  @spec new_http_request(controller) :: {:ok, http_request_id}
+  def new_http_request(controller) do
```
The controller doesn't have a `pid`, it's the request that lives inside a process, so I think you have to rework naming here

noaccOS replied:
In #11 I actually went with [saving the controller's pid](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/11/commits/a0620cd7391f26c6d3ae24f69447daaf167d2dcc#diff-c2f32d20e53f7460e7e01b39b7003c2c8837e763daf382304d668254e4c438a9R30) and [waiting for a response](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/11/commits/a0620cd7391f26c6d3ae24f69447daaf167d2dcc#diff-c2f32d20e53f7460e7e01b39b7003c2c8837e763daf382304d668254e4c438a9R38).

This way we can look at the forwarder as just another function call and leave all the `Plug.Conn` logic inside the controller

rbino commented on lib/edgehog_device_forwarder/connections.ex:
```diff
@@ -0,0 +1,147 @@
+  @spec close_connection(ws_socket_id | http_request_id) :: :ok
+  def close_connection(message_id) do
+    ConCache.delete(@messages, message_id)
+  end
```
I see the same problem that there was in the previous PR: these have to be called explicitly so they don't get cleaned up if, e.g., the request/ws process crashes, leading to resource leaking

noaccOS replied:
**If** the process can crash without calling terminate, then yes. However I'm not yet convinced that is the case

noaccOS replied:
Implemented in #15

noaccOS committed:
feat(cache): http messages

add caching for http messages, which allows
- creating a new request
- forward a message to the cached controller
  - if upgrading, replaces the old entry with the same key
  - if replying, delete the id from the cache

Annopaolo commented on lib/edgehog_device_forwarder/http_requests.ex:
```diff
@@ -0,0 +1,81 @@
+
+          {:error, :http_request_not_found}
+      end
+    end)
```
Consider moving this block (or just the happy path) in a private function in order to improve readability

noaccOS replied:
done

Annopaolo commented on lib/edgehog_device_forwarder/message_queue.ex:
Consider using erlang's `queue`: https://www.erlang.org/doc/man/queue.html
> All operations have an amortized O(1) running time, except [all/2](https://www.erlang.org/doc/man/queue.html#all-2), [any/2](https://www.erlang.org/doc/man/queue.html#any-2), [delete/2](https://www.erlang.org/doc/man/queue.html#delete-2), [delete_r/2](https://www.erlang.org/doc/man/queue.html#delete_r-2), [delete_with/2](https://www.erlang.org/doc/man/queue.html#delete_with-2), [delete_with_r/2](https://www.erlang.org/doc/man/queue.html#delete_with_r-2), [filter/2](https://www.erlang.org/doc/man/queue.html#filter-2), [filtermap/2](https://www.erlang.org/doc/man/queue.html#filtermap-2), [fold/3](https://www.erlang.org/doc/man/queue.html#fold-3), [join/2](https://www.erlang.org/doc/man/queue.html#join-2), [len/1](https://www.erlang.org/doc/man/queue.html#len-1), [member/2](https://www.erlang.org/doc/man/queue.html#member-2), [split/2](https://www.erlang.org/doc/man/queue.html#split-2) that have O(n).

noaccOS replied:
I'm not sure it provides any real benefit over a plain list, as I still need to call the reverse on it at the end

```elixir
queue = :queue.new()

queue = :queue.cons(1, queue)
queue = :queue.cons(2, queue)
queue = :queue.cons(3, queue)
queue = :queue.cons(4, queue)

queue = :queue.to_list(queue)

IO.inspect(queue, pretty: true)
```
```elixir
[4, 3, 2, 1]
```

Annopaolo replied:
That's because you're using what they call "the Okasaki API"
> The "Okasaki API" is inspired by "Purely Functional Data Structures" by Chris Okasaki. It regards queues as lists.

What about the "original" one?
```elixir
q = :queue.new()
q = :queue.in("a", q)  
q = :queue.in("b", q)
q = :queue.in("c", q)
q = :queue.in("d", q)
q = :queue.in("e", q)
q = :queue.in("f", q)
:queue.to_list(q) |> IO.inspect()
# ["a", "b", "c", "d", "e", "f"]
```


(still, :heart: Okasaki :heart:)

noaccOS replied:
Done, replaced `MessageQueue` with `Qex`

Annopaolo commented on test/support/message_case.ex:
```diff
@@ -0,0 +1,55 @@
+        port: 80
+      }
+
+    upgrade =
+      %HTTP.Request{
+        path: "/",
+        method: "GET",
+        query_string: "",
+        headers: %{
+          "content-type" => "plain/text",
+          "status-code" => "101",
+          "upgrade-to" => "websocket"
+        },
+        body: "",
+        port: 80
+      }
```
Nit: consider moving this to the next commit

Annopaolo commented on test/support/message_case.ex:
```diff
@@ -0,0 +1,55 @@
+        body: "contents"
+      }
+
+    upgrade_response =
+      %HTTP.Response{
+        status_code: 101,
+        headers: %{"upgrade" => "websocket"},
+        body: ""
+      }
```
Nit: consider moving this to the next commit

Annopaolo commented on lib/edgehog_device_forwarder/websockets.ex:
```diff
@@ -0,0 +1,166 @@
+
+          {:error, :websocket_not_found}
+      end
+    end)
```
Consider moving this (or even just the happy path) to a private function to improve readability

noaccOS replied:
Done

Annopaolo commented on lib/edgehog_device_forwarder/websockets.ex:
```diff
@@ -0,0 +1,166 @@
+
+          {:error, :websocket_not_found}
+      end
+    end)
```
See above comment

noaccOS replied:
done

Annopaolo commented on test/edgehog_device_forwarder/websockets_test.exs:
```diff
@@ -0,0 +1,134 @@
+  require Logger
+
+  setup_all do
+    [valid_message: {:text, "1"}]
```
This value is constant, so it could be a module attribute rather than a setup_all step

noaccOS replied:
done

Annopaolo commented on test/edgehog_device_forwarder/websockets_test.exs:
```diff
@@ -0,0 +1,134 @@
+    } do
+      assert {:error, :http_request_not_found} ==
+               WebSockets.upgrade(request_id, websocket_socket, data)
+    end
```
Given the error atom, I would expect a different test description

noaccOS replied:
changed the description to `"returns an error if it called without first forwarding an upgrade request"`

Annopaolo commented on test/edgehog_device_forwarder/websockets_test.exs:
```diff
@@ -0,0 +1,134 @@
+      end
+
+      for message <- messages do
+        assert_receive ^message
```
Is the order of messages relevant in this case?

noaccOS replied:
right. fixed in 35d6e4d

noaccOS committed:
chore: add qex dependency

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

Annopaolo commented on lib/edgehog_device_forwarder/websockets/core.ex:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  @spec close_initiate_handshake_from_device(Data.socket_id()) :: :ok
+  def close_initiate_handshake_from_device(socket_id) do
```
This name sounds quite confusing to me: `close` and `initiate` are opposites and are put together as active verbs in the same name. What about this?
```suggestion
  def device_start_socket_closing_handshake(socket_id) do
```

noaccOS replied:
Applied all of them

Annopaolo commented on lib/edgehog_device_forwarder/websockets/core.ex:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  @spec close_initiate_handshake_from_user(Data.socket_id(), user_socket) :: :ok
+  def close_initiate_handshake_from_user(socket_id, device_socket) do
```
See above:
```suggestion
  def user_start_socket_closing_handshake(socket_id, user_socket) do
  ```
  (note the renaming device_socket -> user_socket, too, to match the typespec)

Annopaolo commented on lib/edgehog_device_forwarder/websockets/core.ex:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  @spec close_handshake_complete(Data.socket_id()) :: :ok
+  def close_handshake_complete(socket_id) do
```
See above
```suggestion
  def socket_closing_handshake_complete(socket_id) do
```

Annopaolo commented on lib/edgehog_device_forwarder/websockets/core.ex:
```diff
@@ -0,0 +1,97 @@
+
+  @spec close_initiate_handshake_from_user(Data.socket_id(), user_socket) :: :ok
+  def close_initiate_handshake_from_user(socket_id, device_socket) do
+    "ws: close: socket_id#{inspect(socket_id)}: begin device-requested closing handshake"
```
Shouldn't this be `user-requested`?

rbino commented on lib/edgehog_device_forwarder/http_requests/core.ex:
```diff
@@ -0,0 +1,45 @@
+  def forward(controller, request_id, response) do
+    action = response_action(response)
+
+    case action do
+      :respond ->
+        ConCache.dirty_delete(@table, request_id)
+
+      {:upgrade, :websocket} ->
+        ConCache.dirty_put(@table, request_id, {:upgrading, :websocket, Qex.new()})
+    end
```
Are these `dirty` because you're inside an `isolated` block? In that case you're leaking this implicit assumption inside this separate module, so you should probably move this stuff in the parent module and possibly leave only the pure response generation here.

noaccOS replied:
All ConCache functions in websockets/core and http_requests/core are dirty.

from the websockets.core moduledoc:

>   All of its methods are meant to be used only by the WebSockets module and are meant to be run
  inside an isolated ConCache environment.

right now core/error modules describe what to do in the happy/sad paths, while the parent module performs the checks needed.

rbino commented on lib/edgehog_device_forwarder/http_requests/core.ex:
```diff
@@ -0,0 +1,45 @@
+  defp has_websocket_upgrade_header?(headers) do
+    headers
+    |> Map.get("upgrade", "")
+    |> String.downcase(:ascii)
```
Is this needed? Phoenix should guarantee that you always receive downcased headers

noaccOS replied:
This header map is received from a device response.
We already require the device to send the header name downcased, which was not the default.
IMO downcasing the value ourselves is better.

rbino commented on test/support/message_case.ex:
```diff
@@ -0,0 +1,32 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.MessageCase do
```
Is this used in multiple places? Otherwise moving generation of Plain Old Data in a separate case adds an unnecessary level of indirection

noaccOS replied:
It is used here in `HTTPRequestsTest` and `WebSocketsTest`, and in later PRs it is used inside `ForwarderCase`

noaccOS committed:
feat(cache): add websocket support

the functions allowed are
- finalize upgrade of an http connection to websocket
- forward a message to a socket from its id
  this does not remove the connection as it identifies the socket
- close the socket in two steps, respecting the websockets closing handshake

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>
